Book.java

```diff
@@ -0,0 +1,315 @@
+ *
+ * A Book contains a set of accounts, and is the interface between an agent and its accounts. Agents cannot
+ * directly interact with accounts other than via a Book.
+ *
+ * At the moment, a Book contains an account for each type of contract, plus an equity account and a cash account.
+ *
+ * A simple economic agent will usually have a single Book, whereas complex firms and banks can have several books
+ * (as in branch banking for example).
+ *
+ * @author rafa
+ */
+public class Book implements BookAPI {
+
+    public Book(Agent owner) {
+        // A Book is a list of accounts
+        accounts = new HashSet<>();
+
+        // Subsets of the list of accounts (for quicker searching)
+        assetAccounts = new HashSet<>();
+        liabilityAccounts = new HashSet<>();
+        equityAccounts = new HashSet<>();
+
+        // Each Account includes an inventory to hold one type of contract.
+        // These hashmaps are used to access the correct account for a given type of contract.
+        // Note that separate hashmaps are needed for asset accounts and liability accounts: the same contract
+        // type (such as Loan) can sometimes be an asset and sometimes a liability.
+        assetAccountsMap = new HashMap<>();
+        liabilityAccountsMap = new HashMap<>();
+
+        // Not sure whether I should be passing the owner...
+        this.owner = owner;
+
+        // A book is initially created with a cash account and an equityAccounts account (it's the simplest possible book)
+        cashAccount = new Account("cash", AccountType.ASSET);
+        equityAccount = new Account("equityAccounts", AccountType.EQUITY);
+        addAccount(cashAccount, null);
+        addAccount(equityAccount, null);
+    }
+
+
+    private Agent owner;
+    private HashSet<Account> accounts;
+    private HashSet<Account> assetAccounts;
+    private HashSet<Account> liabilityAccounts;
+    private HashSet<Account> equityAccounts;
+    private HashMap<Class<? extends contracts.Contract>, Account> assetAccountsMap;
+    private HashMap<Class<? extends contracts.Contract>, Account> liabilityAccountsMap;
+    private Account cashAccount;
+    private Account equityAccount;
+
+    public double getAssetValue() {
+        double assetTotal = 0;
+        for (Account assetAccount : assetAccounts) {
+            assetTotal+=assetAccount.getBalance();
+        }
+        return assetTotal;
+    }
+
+    public double getLiabilityValue() {
+        double liabilityTotal = 0;
+        for (Account liabilityAccount : liabilityAccounts) {
+            liabilityTotal+=liabilityAccount.getBalance();
+        }
+        return liabilityTotal;
+    }
+
+    public double getEquityValue() {
+        double equityTotal = 0;
+        for (Account equityAccount : equityAccounts) {
+            equityTotal += equityAccount.getBalance();
+        }
+        return equityTotal;
+    }
+
+    public double getAssetValueOf(Class<? extends Contract> contractType) {
```

Comment:
I'm a little confused. How are contracts being valued? Can anyone send in a subjective function for evaluation?